# Imports

In [121]:
import pandas as pd

from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

from collections import Counter

# Dataset Inspection

In [2]:
# Load the cleaned data
df = pd.read_csv('clean_df.csv')

# Display the first few rows of the DataFrame to verify it's loaded correctly
df.head(20)

quantity  unit                         ingredient  ingredient_step  \
0       45.0    ml                           dark rum                0   
1       22.5    ml                         lime juice                1   
2       15.0    ml                              sugar                2   
3        1.0  dash                          angostura                3   
4        6.0  drop                             pernod                4   
5      240.0    ml                        crushed ice                5   
6       60.0    ml                     silver tequila                0   
7       22.5    ml       marie brizard creme de cacao                1   
8       22.5    ml                        lemon juice                2   
9       45.0    ml                                gin                0   
10      22.5    ml  mari brizard white creme de cacao                1   
11      22.5    ml                       lillet blanc                2   
12      22.5    ml                        lemon juice                3   
13      45.0    ml                             pernod                0   
14       1.0   NaN                              sugar                1   
15     150.0    ml                              water                2   
16      30.0    ml                       gold tequila                0   
17      30.0    ml                           gold rum                1   
18      60.0    ml                   grapefruit juice                2   
19      90.0    ml                    pineapple juice                3   

    recipe_id  
0           0  
1           0  
2           0  
3           0  
4           0  
5           0  
6           1  
7           1  
8           1  
9           2  
10          2  
11          2  
12          2  
13          3  
14          3  
15          3  
16          4  
17          4  
18          4  
19          4

In [3]:
df.tail(5)

quantity  unit       ingredient  ingredient_step  recipe_id
2887      60.0    ml      rye whiskey                0        654
2888      22.5    ml      lemon juice                1        654
2889      22.5    ml  pineapple syrup                2        654
2890      15.0    ml      dolin blanc                3        654
2891       1.0  dash   orange bitters                4        654

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2892 entries, 0 to 2891
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   quantity         2586 non-null   float64
 1   unit             2524 non-null   object 
 2   ingredient       2890 non-null   object 
 3   ingredient_step  2892 non-null   int64  
 4   recipe_id        2892 non-null   int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 113.1+ KB


In [5]:
# Group by recipe_id and aggregate ingredients into lists
grouped_df = df.groupby('recipe_id')['ingredient'].apply(list).reset_index()

# Display grouped DataFrame
print(grouped_df)

     recipe_id                                         ingredient
0            0  [dark rum, lime juice, sugar, angostura, perno...
1            1  [silver tequila, marie brizard creme de cacao,...
2            2  [gin, mari brizard white creme de cacao, lille...
3            3                             [pernod, sugar, water]
4            4  [gold tequila, gold rum, grapefruit juice, pin...
..         ...                                                ...
650        650      [light rum, galliano, triple sec, lime juice]
651        651  [dark rum, light rum, tia maria, orange juice,...
652        652     [vodka, galliano, lime juice, pineapple juice]
653        653        [yellow chartreuse, pernod, apricot brandy]
654        654  [rye whiskey, lemon juice, pineapple syrup, do...

[655 rows x 2 columns]


In [6]:
# Extract the list of lists of ingredients
recipes = grouped_df['ingredient'].tolist()

# Get the list of all unique ingredients
all_ingredients = df['ingredient'].unique().tolist()

# Use MultiLabelBinarizer to encode the ingredients
mlb = MultiLabelBinarizer(classes=all_ingredients)
encoded_recipes = mlb.fit_transform(recipes)

# Create a DataFrame for easier manipulation
encoded_df = pd.DataFrame(encoded_recipes, columns=mlb.classes_)

In [7]:
encoded_df.shape

(655, 389)

# Outline:

Model Building:

Our model architecture is:

l0: 389 (amount of unique ingredients)

hidden layers:          relu

L :            sigmoid   , 389

In [9]:
input_dim = len(all_ingredients)  # Number of unique ingredients

model = Sequential()
model.add(Dense(units=128, input_dim=input_dim, activation='relu'))
model.add(Dense(units=128, activation='relu'))
model.add(Dense(units=input_dim, activation='sigmoid'))  # Output layer with sigmoid activation

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [39]:
# attempt at at a 2nd model, larger architecture since 389 param

input_dim = len(all_ingredients)  # Number of unique ingredients

model = Sequential()
model.add(Dense(units=256, input_dim=input_dim, activation='relu'))  # First hidden layer
model.add(Dense(units=128, activation='relu'))  # Second hidden layer
model.add(Dense(units=256, activation='relu'))  # Third hidden layer
model.add(Dense(units=input_dim, activation='sigmoid'))  # Output layer

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Print model summary to verify the architecture
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 256)               99840     
                                                                 
 dense_4 (Dense)             (None, 128)               32896     
                                                                 
 dense_5 (Dense)             (None, 256)               33024     
                                                                 
 dense_6 (Dense)             (None, 389)               99973     
                                                                 
Total params: 265,733
Trainable params: 265,733
Non-trainable params: 0
_________________________________________________________________


Model Training:

In [48]:
X = encoded_df.values  # Input features (binary vectors of ingredients)
y = encoded_df.values  # Target labels (same as input for multi-label classification)

model.fit(X, y, epochs=1, batch_size=50, validation_split=0.2)

11/11 [==============================] - 0s 8ms/step - loss: 2.9224e-04 - accuracy: 0.5553 - val_loss: 0.0702 - val_accuracy: 0.5802


Generating New Recipes:

Start with a seed ingredient or set of ingredients.
Use the trained model to predict additional ingredients.

In [123]:
model.fit(X, y, epochs=1, batch_size=50, validation_split=0.2)

seed_ingredients = ['vodka']  # Example seed ingredients
seed_vector = mlb.transform([seed_ingredients])[0]

predicted_probabilities = model.predict(seed_vector.reshape(1, -1))
predicted_ingredients = mlb.inverse_transform((predicted_probabilities > 0.00001).astype(int))[0]

generated_recipe = list(set(seed_ingredients + list(predicted_ingredients)))
print("Generated Recipe:", generated_recipe)


1/1 [==============================] - 0s 14ms/step
Generated Recipe: ['dry sake', 'ginger beer', 'vodka', 'plum wine', 'kahlúa', 'lemon wedge']


In [125]:
import pandas as pd
from collections import Counter

# Number of times to generate recipes
num_iterations = 100

# Example seed ingredients
seed_ingredients = ['vodka']
seed_vector = mlb.transform([seed_ingredients])[0]

# List to store generated recipes
generated_recipes = []

for _ in range(num_iterations):
    # Fit the model for one epoch
    model.fit(X, y, epochs=1, batch_size=50, validation_split=0.2, verbose=0)
    
    # Predict probabilities
    predicted_probabilities = model.predict(seed_vector.reshape(1, -1))
    
    # Convert probabilities to ingredients
    predicted_ingredients = mlb.inverse_transform((predicted_probabilities > 0.00001).astype(int))[0]
    
    # Generate recipe
    generated_recipe = list(set(seed_ingredients + list(predicted_ingredients)))
    
    # Store the generated recipe as a tuple for easier counting
    generated_recipes.append(tuple(generated_recipe))

# Count the occurrences of each unique recipe
recipe_counts = Counter(generated_recipes)

# Convert to DataFrame
df = pd.DataFrame(recipe_counts.items(), columns=['Recipe', 'Count'])

# Sort by count (optional)
df = df.sort_values(by='Count', ascending=False)

print(df)


1/1 [==============================] - 0s 15ms/step
                                              Recipe  Count
2          (dry sake, vodka, plum wine, lemon wedge)     65
1  (dry sake, vodka, plum wine, kahlúa, lemon wedge)     27
0  (dry sake, ginger beer, vodka, plum wine, kahl...      8


In [129]:
df['Recipe']

2            (dry sake, vodka, plum wine, lemon wedge)
1    (dry sake, vodka, plum wine, kahlúa, lemon wedge)
0    (dry sake, ginger beer, vodka, plum wine, kahl...
Name: Recipe, dtype: object

In [128]:
df['Recipe'][0]

('dry sake', 'ginger beer', 'vodka', 'plum wine', 'kahlúa', 'lemon wedge')

In [19]:
predicted_probabilities

array([[3.95206846e-02, 2.13554013e-03, 1.06728841e-02, 1.17316877e-03,
        1.10128392e-02, 1.99002889e-03, 1.05354413e-02, 4.67673643e-03,
        3.47359432e-03, 9.24138259e-03, 5.45110379e-04, 2.92723160e-03,
        3.45947733e-03, 1.91824157e-02, 1.03301061e-02, 6.56192750e-02,
        1.55216344e-02, 2.32320987e-02, 2.68419040e-03, 1.29588961e-03,
        2.07096091e-04, 1.54185640e-02, 3.41800245e-04, 1.59413833e-02,
        2.42683161e-02, 1.13276355e-02, 2.68909223e-02, 2.16340576e-03,
        2.01898646e-02, 4.03897138e-03, 1.53216124e-02, 5.28924586e-03,
        1.31054120e-02, 2.16253707e-03, 2.67793704e-02, 2.49111233e-03,
        2.98502506e-03, 2.43437216e-02, 1.00448452e-01, 1.57015771e-02,
        2.06285790e-02, 4.60406579e-02, 4.59817052e-02, 5.04731492e-04,
        1.21376766e-02, 4.88650007e-03, 1.18583441e-02, 1.11335507e-02,
        8.67285300e-03, 1.08531676e-02, 2.81302836e-02, 1.21530024e-02,
        2.46498297e-04, 4.42771567e-03, 4.41456214e-02, 1.868071

Summary of Steps:
Clean Data Preparation:

Create a DataFrame with binary encoded ingredient vectors for each recipe.
Model Building:

Define a neural network to learn from these vectors.
Model Training:

Train the model on your dataset.
Recipe Generation:

Generate new recipes starting from seed ingredients.
By following these steps, you should be able to create and train a model that can generate new cocktail recipes based on the patterns it learns from your dataset. Let me know if you need more detail on any of these steps!

The architecture described here is a simple feedforward neural network, also known as a Multi-Layer Perceptron (MLP). This architecture is used for multi-label classification, where each ingredient is treated as a binary label that can be present or absent in a recipe.

Architecture Explanation
Input Layer:

The input layer consists of nodes equal to the number of unique ingredients. Each node represents whether a particular ingredient is present (1) or absent (0) in the recipe.
Input Dimension: input_dim = len(all_ingredients).
Hidden Layers:

The network has two hidden layers, each with 128 neurons. These layers use the ReLU (Rectified Linear Unit) activation function, which introduces non-linearity to the model and allows it to learn complex patterns.
First Hidden Layer: Dense(units=128, input_dim=input_dim, activation='relu')
Second Hidden Layer: Dense(units=128, activation='relu')
Output Layer:

The output layer also consists of nodes equal to the number of unique ingredients. Each node represents the probability of the corresponding ingredient being part of the recipe.
The sigmoid activation function is used in the output layer to produce probabilities between 0 and 1 for each ingredient.
Output Layer: Dense(units=input_dim, activation='sigmoid')
Loss Function:

The model uses binary cross-entropy loss, suitable for multi-label classification where each label (ingredient) is a binary decision (present or absent).
Loss Function: binary_crossentropy
Optimizer:

The Adam optimizer is used to minimize the loss function during training. Adam is a popular choice due to its adaptive learning rate and efficiency.
Optimizer: adam
Metrics:

The accuracy metric is used to evaluate the model's performance during training and validation.
Metrics: accuracy
Why This Architecture?
Simplicity: This architecture is straightforward and easy to implement, making it suitable for a dataset of your size (654 recipes and 447 unique ingredients).
Flexibility: The use of dense layers allows the network to learn from the presence and absence of each ingredient, capturing the relationships between different ingredients.
Scalability: Adding more hidden layers or increasing the number of neurons in each layer can improve the model's capacity to learn more complex patterns, if needed.
Summary
This architecture is a feedforward neural network designed for multi-label classification. It takes a binary vector representing the presence or absence of each ingredient as input and outputs a binary vector representing the predicted probabilities of each ingredient being part of the cocktail recipe. This approach is suitable for generating new recipes by learning the common ingredient combinations from the training data.






